packages/app/src/systems/DApp/machines/txApproveMachine.tsx

```diff
@@ -0,0 +1,198 @@
+import type {
+  TransactionRequest,
+  TransactionResponse,
+  TransactionResultReceipt,
+  Wallet,
+} from 'fuels';
+import type { InterpreterFrom, StateFrom } from 'xstate';
+import { assign, createMachine } from 'xstate';
+import { send } from 'xstate/lib/actions';
+
+import type { UnlockMachine, UnlockMachineEvents } from './unlockMachine';
+import { unlockMachine } from './unlockMachine';
+
+import type { AccountInputs } from '~/systems/Account';
+import type { ChildrenMachine } from '~/systems/Core';
+import { FetchMachine, provider } from '~/systems/Core';
+import type { VMApiError } from '~/systems/Transaction';
+
+type MachineContext = {
+  tx?: TransactionRequest;
+  receipts?: TransactionResultReceipt[];
+  approvedTx?: TransactionResponse;
+  txDryRunError?: VMApiError;
+};
+
+type MachineServices = {
+  approveTx: {
+    data: TransactionResponse;
+  };
+  calculateGas: {
+    data: TransactionResultReceipt[];
+  };
+};
+
+type MachineEvents =
+  | { type: 'UNLOCK_WALLET'; input: AccountInputs['unlock'] }
+  | { type: 'START_APPROVE'; input?: null }
+  | { type: 'CALCULATE_GAS'; input: { tx: TransactionRequest } }
+  | { type: 'CLOSE_UNLOCK'; input?: null };
+
+export const txApproveMachine = createMachine(
+  {
+    // eslint-disable-next-line @typescript-eslint/consistent-type-imports
+    tsTypes: {} as import('./txApproveMachine.typegen').Typegen0,
+    schema: {
+      context: {} as MachineContext,
+      services: {} as MachineServices,
+      events: {} as MachineEvents,
+    },
+    predictableActionArguments: true,
+    id: '(machine)',
+    initial: 'waitingTxRequest',
+    states: {
+      waitingTxRequest: {
+        on: {
+          CALCULATE_GAS: {
+            target: 'calculatingGas',
+            actions: ['assignTx'],
+          },
+        },
+      },
+      calculatingGas: {
+        invoke: {
+          src: 'calculateGas',
+          data: (_: MachineContext) => ({ input: { tx: _.tx } }),
+          onDone: [
+            {
```

Comment:
It's possible to use the FetchMachine.errorState it should target failed maybe. Also, it already auto assign on the parent machine the field error with the error message. Removing the requirement of having the `assignTxDryRunError`

Reply:
if you have more than one `fetchMachine` and wants to track the errors in different context variables, will not work.

that's why I prefer to use on this way.

but we should refact and move to `onFailed` as discussed before in slack channel

Reply:
I think we should use the common field error and the `FetchMachine.errorState` to have a common pattern and we can move to onFailed after is refactored